File: CSharp/Eta/Applications/Examples/WatchlistConsumer/ChannelInfo.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license      --
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
 *|                See the project's LICENSE.Md for details.                  --
 *|           Copyright (C) 2023 Refinitiv. All rights reserved.         --
 *|-----------------------------------------------------------------------------
 */

namespace LSEG.Eta.ValueAdd.WatchlistConsumer;

using LSEG.Eta.Codec;
using LSEG.Eta.Example.VACommon;
using LSEG.Eta.Transports;
using LSEG.Eta.ValueAdd.Rdm;
using LSEG.Eta.ValueAdd.Reactor;

/// <summary>
/// Contains information associated with each open channel in the Watchlist Consumer.
/// </summary>
internal class ChannelInfo
{
    public ReactorConnectOptions ConnectOptions { get; set; } = new();
    public ReactorConnectInfo ConnectInfo { get; set; } = new();
    public ConsumerRole ConsumerRole { get; set; } = new();
    public PostHandler PostHandler { get; set; } = new();
    public DataDictionary Dictionary { get; set; } = new();
    public int FieldDictionaryStreamId { get; set; }
    public int EnumDictionaryStreamId { get; set; }
    public bool ShouldOffStreamPost { get; set; }
    public bool ShouldOnStreamPost { get; set; }
    public bool ShouldEnableEncrypted { get; set; }
    public EncryptionProtocolFlags EncryptionProtocol { get; set; }
    public Buffer PostItemName { get; set; } = new();
    public DecodeIterator DIter { get; set; } = new();
    public Msg ResponseMsg { get; set; } = new();
    public LoginRefresh LoginRefresh { get; set; } = new();
    public bool HasServiceInfo { get; set; }
    public Service ServiceInfo { get; set; } = new();
    public ReactorChannel? ReactorChannel { get; set; }

    /// <summary>
    /// Represented by epoch time in milliseconds.
    /// </summary>
    public System.DateTime LoginReissueTime { get; set; }

    public ConnectionArg ConnectionArg { get; set; } = new();
    public bool CanSendLoginReissue { get; set; }

    public ChannelInfo()
    {
        ConnectOptions.ConnectionList.Add(ConnectInfo);
    }
}

File: Java/Eta/Applications/Examples/src/main/java/com/refinitiv/eta/valueadd/examples/watchlistconsumer/ChannelInfo.java
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2020,2022,2024 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

package com.refinitiv.eta.valueadd.examples.watchlistconsumer;

import com.refinitiv.eta.codec.Buffer;
import com.refinitiv.eta.codec.CodecFactory;
import com.refinitiv.eta.codec.DataDictionary;
import com.refinitiv.eta.codec.DecodeIterator;
import com.refinitiv.eta.codec.Msg;
import com.refinitiv.eta.valueadd.domainrep.rdm.directory.DirectoryMsgFactory;
import com.refinitiv.eta.valueadd.domainrep.rdm.directory.Service;
import com.refinitiv.eta.valueadd.domainrep.rdm.login.LoginMsgFactory;
import com.refinitiv.eta.valueadd.domainrep.rdm.login.LoginMsgType;
import com.refinitiv.eta.valueadd.domainrep.rdm.login.LoginRefresh;
import com.refinitiv.eta.valueadd.examples.common.ConnectionArg;
import com.refinitiv.eta.valueadd.reactor.ConsumerRole;
import com.refinitiv.eta.valueadd.reactor.ReactorChannel;
import com.refinitiv.eta.valueadd.reactor.ReactorConnectInfo;
import com.refinitiv.eta.valueadd.reactor.ReactorConnectOptions;
import com.refinitiv.eta.valueadd.reactor.ReactorFactory;
import com.refinitiv.eta.valueadd.reactor.TunnelStream;

/*
 * Contains information associated with each open channel
 * in the value add Consumer.
 */
class ChannelInfo
{
	ConnectionArg connectionArg;
    ReactorConnectOptions connectOptions = ReactorFactory.createReactorConnectOptions();
    ReactorConnectInfo connectInfo = ReactorFactory.createReactorConnectInfo();
    ConsumerRole consumerRole = ReactorFactory.createConsumerRole();

    PostHandler postHandler = new PostHandler();
 
    DataDictionary dictionary;
    int fieldDictionaryStreamId = 0;
    int enumDictionaryStreamId = 0;
    boolean shouldOffStreamPost = false;
    boolean shouldOnStreamPost = false;
    boolean shouldEnableEncrypted = false;
    boolean shouldEnableHttp = false;
    Buffer postItemName = CodecFactory.createBuffer();
    
    DecodeIterator dIter = CodecFactory.createDecodeIterator();
    Msg responseMsg = CodecFactory.createMsg();

    LoginRefresh loginRefresh = (LoginRefresh)LoginMsgFactory.createMsg();
	boolean hasServiceInfo = false;
    Service serviceInfo = DirectoryMsgFactory.createService();
    ReactorChannel reactorChannel;
    
    boolean tunnelStreamOpenSent = false; // flag to track if we already made a tunnel stream open request
    
    // assume one tunnel stream per ReactorChannel
    TunnelStream tunnelStream;
    boolean hasTunnelStreamServiceInfo = false;
    Service tsServiceInfo = DirectoryMsgFactory.createService();
    boolean isTunnelStreamUp;
    
	long loginReissueTime; // represented by epoch time in milliseconds
	boolean canSendLoginReissue;

	ChannelInfo()
    {
    	connectOptions.connectionList().add(connectInfo);
    	loginRefresh.rdmMsgType(LoginMsgType.REFRESH);
    }
}
